pkg/policy/distillery/distillery.go

```diff
@@ -0,0 +1,184 @@
+// if the cached policy is already at the revision specified in the repo.
+//
+// Returns whether the cache was updated, or an error.
+//
+// Must be called with repo.Mutex held for reading.
+func (cache *policyCache) updateSelectorPolicy(repo PolicyRepository, ep Endpoint) (bool, error) {
+	identity := ep.GetSecurityIdentity()
+	revision := repo.GetRevision()
+
+	// Don't resolve policy if it was already done for this Identity.
+	var currentRevision uint64
+	cache.Lock()
+	cip, ok := cache.policies[identity]
+	if ok {
+		currentRevision = cip.revision
+	}
+	cache.Unlock()
+	if !ok {
+		return false, fmt.Errorf("SelectorPolicy not found in cache for ID %d", identity.ID)
+	} else if revision == currentRevision {
+		return false, nil
+	}
+
+	// Resolve the policies, which could fail
+	identityPolicy, err := repo.ResolvePolicyLocked(identity)
+	if err != nil {
+		return false, err
+	}
+
+	// We don't cover the ResolvePolicyLocked() call above with the cache
+	// Mutex because it's potentially expensive, and endpoints with
+	// different identities should be able to concurrently compute policy.
+	//
+	// However, as long as UpdatePolicy() is triggered from endpoint
+	// regeneration, it's possible for two endpoints with the *same*
+	// identity to race to the revision check above, both find that the
+	// policy is out-of-date, and resolve the policy then race down to
+	// here. Don't update the policy if we're late to the party.
+	//
+	// Note that because repo.Mutex is held, the two racing threads will be
+	// guaranteed to compute policy for the same revision of the policy.
+	// We could save some CPU by, for example, forcing resolution of policy
+	// for the same identity to block on a channel/lock, but this is
+	// skipped for now as there are upcoming changes to the cache update
+	// logic which would render such mechanisms obsolete.
+	cache.Lock()
+	defer cache.Unlock()
+	changed := revision > cip.revision
+	if changed {
+		cip.setPolicyLocked(identityPolicy, revision)
+	}
+	return changed, nil
+}
+
+// Upsert notifies the global policy cache that the specified endpoint requires
+// a reference to an identity policy, and returns the cached identity policy
+// for that identity.
+func Upsert(ep Endpoint) SelectorPolicy {
+	identity := ep.GetSecurityIdentity()
+	cip, _ := globalPolicyCache.upsert(identity, ep)
+	return cip
+}
+
+// Remove a cached SelectorPolicy reference for the specified endpoint.
+func Remove(ep Endpoint) error {
+	_, err := globalPolicyCache.remove(ep)
+	return err
+}
+
+// PolicyRepository is an interface which generates an SelectorPolicy for a
+// particular Identity.
+type PolicyRepository interface {
+	ResolvePolicyLocked(*identityPkg.Identity) (*policy.SelectorPolicy, error)
+	GetRevision() uint64
+}
+
+// Endpoint represents a user of an SelectorPolicy. It is used for managing
+// the lifetime of the cached policy.
+type Endpoint interface {
```

Comment:
IMO "Owner" would be better name for this than "Endpoint", especially when this is NOT `endpoint.Endpoint`

Reply:
It is currently, though. We don't want it to be longer term, but right now it is:

```
$ git grep "distillery\.[Upsert|Remove]"                                                                                                                                                                                                       
pkg/endpoint/endpoint.go:               if err := distillery.Remove(e); err != nil {
pkg/endpoint/policy.go:         e.SelectorPolicy = distillery.Upsert(e)
pkg/endpoint/policy.go: if err := distillery.UpdatePolicy(repo, e); err != nil {
```

Reply:
By the way I'd like to trim this back from using `policy.PolicyOwner` to something simpler, but this is what's required for now as an incremental step.

Reply:
I think I can add an incremental patch on top of this to switch it over already, actually. Let me try that.

Reply:
Ah, so this is how the reference-counting is done currently. Basically the same SelectorPolicy can be found by two different endpoints, then the endpoint pointer is added to the list of `users`, which allows the distillery to avoid deleting a SelectorPolicy if it's still in use.

If we switch this over to use the identitymanager then it will no longer be necessary. However this will also require rearranging the lifecycle of the SelectorPolicy cache. I'd like to handle that in a separate PR.